### PR TITLE
Added vocab_restrictor

### PR DESCRIPTION
**Pull request by sandoram**

### CODE REVIEW

**Comment by mille-s** [comment]
Hi, thanks for the original transformation!  Does the system try to replace noun phrases only? How do you indicate which words to replace and which not?

I suppose this tranformation could be useful in some tasks that take texts as input, but is probably more suited to produce additional training data. And people can also just play with it and have fun :)

**Review by mille-s — APPROVED** [review]

**Comment by kaustubhdhole** [comment]
@sandoram , please pull main and check since there is a build failure in your changes. Once the build is fixed, I am happy to merge.

**Review by vgtomahawk — APPROVED** [review]
I think this is quite a novel and interesting transformation / data augmentation Technique by @sandoram et al - it would be great if the authors could just try to resolve and get over the build failure...

File: transformations/vocab_restrictor/grammarize.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
+import spacy
+import pattern3
+import re
+import lemminflect
+import nltk
+nltk.download('wordnet')
+from nltk.corpus import wordnet as wn
+nlp = spacy.load("en_core_web_sm")
```
Comment (kaustubhdhole) [inline transformations/vocab_restrictor/grammarize.py:8]: Please use the global spacy.

File: transformations/vocab_restrictor/transformation.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    target_words = [normalize(word,vocab,tree) for word in words]
+    restricted_sentence = ' '.join([t if t==w else grammarize(w, t)
+                         for t,w in zip(target_words,words)])   
+    return restricted_sentence
+    
+class VocabRestrictor(SentenceOperation):
+    tasks = [
+        TaskType.TEXT_CLASSIFICATION,
+        TaskType.TEXT_TO_TEXT_GENERATION,
+    ]
```
Comment (kaustubhdhole) [inline transformations/vocab_restrictor/transformation.py:181]: Please add keywords here.

File: transformations/vocab_restrictor/transformation.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+import spacy
+import lemminflect
+from interfaces.SentenceOperation import SentenceOperation
+from tasks.TaskTypes import TaskType
+import nltk
+nltk.download('wordnet')
+from nltk.corpus import wordnet as wn
+
+nlp = spacy.load("en_core_web_sm") 
+module_use = hub.load("https://tfhub.dev/google/universal-sentence-encoder-multilingual-qa/3")
```
Comment (kaustubhdhole) [inline transformations/vocab_restrictor/transformation.py:18]: Let spacy and USE go into the constructor.

File: transformations/vocab_restrictor/requirements.txt [hunk 4]
```diff
@@ -0,0 +1,10 @@
+tensorflow==2.5.0
+tensorflow_hub==0.12.0
+tensorflow_text==2.5.0
+scipy==1.7.1
+pattern3
+lemminflect==0.2.2
+nltk==3.2.5
+pandas==1.1.5
+spacy==3.0.0
+https://github.com/explosion/spacy-models/releases/download/en_core_web_sm-3.0.0/en_core_web_sm-3.0.0.tar.gz
```
Comment (kaustubhdhole) [inline transformations/vocab_restrictor/requirements.txt:10]: Most of these requirements are already present in the main requirements.txt. So, please remove this requirements.txt file.

File: transformations/vocab_restrictor/test.json [hunk 5]
```diff
@@ -0,0 +1 @@
...rld in all its unaskeded glory, and still to love it. -Oscar Wilde"]}}, {"class": "VocabRestrictor", "inputs": {"sentence": "If life were predictable it would cease to be life, and be without flavor. -Eleanor Roosevelt"}, "outputs": {"sentence": ["If life were opportune it would cease to be life, and be without relish. -Ariel Roosevelt", "If life were preposterously it would cease to be life, and be without fragrance. -Lily Roosevelt", "If life were forcible it would cease to be life, and be without bitter. -Owenite Roosevelt", "If life were forcible it would cease to be life, and be without flavour. -\u0152chalian Roosevelt", "If life were propitious it would cease to be life, and be without flavor. -Elne Roosevelt", "If life were expectantly it would stop to be life, and be without flavor. -Lawrence Roosevelt", "If life were uncertain it would cease to be life, and be without flavor. -Elengi Roosevelt", "If life were hopeful it would stop to be life, and be without flavour. -Elsie Roosevelt", "If life were improbable it would cease to be life, and be without taste. -Llyn Roosevelt", "If life were unexpectedly it would cease to be life, and be without taste. -Eliza Roosevelt"]}}]}
```
Comment (kaustubhdhole) [inline transformations/vocab_restrictor/test.json:1]: Also you might want to reformat this json so that is clear to everyone reading it.

**Comment by kaustubhdhole** [comment]
Also seems the test cases are failing so this can't be merged until then. Please fix the test cases. Sorry to be the bad guy here :)

**Comment by kaustubhdhole** [comment]
@sandoram ping!

**Comment by kaustubhdhole** [comment]
Feel free to open it again!

**Pull request closed without merging**